web/modules/custom/openideal_challenge/src/Service/OpenidealChallengeService.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+namespace Drupal\openideal_challenge\Service;
+
+use Drupal\Component\Datetime\TimeInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\node\NodeInterface;
+use Drupal\openideal_challenge\Event\OpenidealChallengeEvent;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * Class OpenidealChallengeService.
+ *
+ * @package Drupal\openideal_challenge\Service
+ */
+class OpenidealChallengeService implements OpenidealChallengeServiceInterface {
+
+  /**
+   * The count of nodes which will be processed at a time.
+   */
+  const LIMIT = 5;
+
+  /**
+   * Entity Type Manager service object.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The event dispatcher.
+   *
+   * @var \Symfony\Component\EventDispatcher\EventDispatcherInterface
+   */
+  protected $eventDispatcher;
+
+  /**
+   * The time service.
+   *
+   * @var \Drupal\Component\Datetime\TimeInterface
+   */
+  protected $time;
+
+  /**
+   * Constructs a OpenidealChallengeService object.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager.
+   * @param \Symfony\Component\EventDispatcher\EventDispatcherInterface $eventDispatcher
+   *   An event dispatcher instance to use for configuration events.
+   * @param \Drupal\Component\Datetime\TimeInterface $time
+   *   The time service.
+   */
+  public function __construct(EntityTypeManagerInterface $entityTypeManager, EventDispatcherInterface $eventDispatcher, TimeInterface $time) {
+    $this->entityTypeManager = $entityTypeManager;
+    $this->eventDispatcher = $eventDispatcher;
+    $this->time = $time;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function challengeOpenCloseOperation($operation = 'open') {
+    $result = FALSE;
+    $operation_field_name = $operation == 'open' ? 'field_open_on' : 'field_close_on';
+    $event_name = $operation == 'open' ? OpenidealChallengeEvent::CHALLENGE_OPEN : OpenidealChallengeEvent::CHALLENGE_CLOSE;
+
+    // Select all nodes of the challenge type that are enabled for scheduled
+    // closing/opening and where close_on/open_on is less than or equal
+    // to the current time.
+    $query = $this->entityTypeManager->getStorage('node')->getQuery()
+      ->exists($operation_field_name)
+      ->condition($operation_field_name, $this->time->getRequestTime(), '<=')
+      ->condition('type', 'challenge', '=')
+      ->condition('status', NodeInterface::PUBLISHED)
+      ->latestRevision()
+      ->sort($operation_field_name)
+      ->sort('nid')
+      ->range(0, self::LIMIT);
+    $nids = $query->execute();
+
+    if (empty($nids)) {
+      return $result;
+    }
+
+    $nodes = $this->entityTypeManager->getStorage('node')
+      ->loadMultiple($nids);
+    /** @var \Drupal\node\NodeInterface $node */
+    foreach ($nodes as $node) {
+      if ($node->hasField('field_is_open')) {
```

Comment:
Should we move this condition to the query above?
If, for some reason, there is no field "Is Open", then we should probably ignore the whole process.